### ISSUE TITLE
firebase_auth: verifyPhoneNumber should not sign in

### ISSUE THREAD

**Issue by kdy1** [flutter/flutter#22304]
I want to link phone number to my user's account.

But I can't as `verifyPhoneNumber` implicitly signs in on native side.
https://github.com/flutter/plugins/blob/6341d3a5c7b36194ab0fd50a839b363579c3eb37/packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java#L161-L176

To be consistent with other methods, it should instead return Future<`sms code`>, and report `verificationId` with callback as it might be expired while waiting for the code sent.

```dart
/// 
Future<String> verifyPhoneNumber({
  @required String phoneNumber,
  @required Duration timeout,
  int forceResendingToken,
  @required PhoneCodeSent codeSent,
  @required PhoneCodeAutoRetrievalTimeout codeAutoRetrievalTimeout,
}) async {}

Future<FirebaseUser> signInWithPhoneNumber({
  @required String verificationId,
  @required String smsCode,
}) async {}

Future<FirebaseUser> linkWithPhoneNumber({
  @required String verificationId,
  @required String smsCode,
}) async {
```

But this is breaking change.

**Comment by creativecreatorormaybenot** [comment]
@kdy1 `verifyPhoneNumber` does not sign in. You can easily test that. The `verificationId` can be accessed through the `codeSent` callback.  
I am not sure when `onVerificationCompleted` is called, but definitely not in the process of receiving the sms code. You can use either [`linkWithPhoneNumber`](https://github.com/flutter/plugins/pull/809#issuecomment-427263454) or `signInWithPhoneNumber` after sending the code.  

It might be possible that this happens when you are *signed out*, but I do not think so. I only tested it with anonymous accounts, but either way: If you want to link the account, *verifyPhoneNumber* **will not** sign you in with a different account and if you want to use `signInWithPhoneNumber`, it does not matter if it **would be** called natively because you would want to call it anyway.  

However, I am quite confused with that implementation in general. I am pretty sure that `onVerificationCompleted` is never called in the process (you can prove me wrong, I only tested it with anonymous accounts) because it would also not make any sense. In order for an account to be created via a phone number, it needs to be verified in the first place, which is why I do not understand when `onVerificationCompleted` is thought to be executed in the lifecycle.

### TL;DR
I believe that the current implementation is totally fine in the regard you mentioned and that the problem does not exist.

**Comment by kdy1** [comment]
But if you authenticate user then call verifyPhoneNumber, user account is switched.

I digged the issue, and native code calls `signInWithCredential`. I think that's verifySmsCode accepts `onVerficationComplete(FirebaseUser user)`.

**Comment by creativecreatorormaybenot** [comment]
@kdy1 But not before verifying?  

Signing in before verifying cannot work because you need both the `verificationId` and `smsCode` to sign in with a phone number and the sms code is **isolated** from the *application*.

**Comment by gildaswise** [comment]
@creativecreatorormaybenot I get the problem, @kdy1 is saying that the only way to verify the **SMS code** is to use `signInWithPhoneNumber`, and that switches with whatever account is currently signed in. IIRC, on Android, there's a separate method for just verifying the SMS code, and that would solve the problem. I don't know on iOS but I am also looking for a solution to this as it breaks the signin/signup flow for many apps.

**Comment by kdy1** [comment]
@creativecreatorormaybenot  `onVerificationCompleted(PhoneAuthCredential)` might be called automatically.

https://firebase.google.com/docs/auth/android/phone-auth?hl=en#verification-callbacks


And when it's called automatically, current api does not provide any way to get the sms code. (Instead it signs in)
`timeout` parameter controls only the duration of automatic retrieval. You can check this by setting it larger than 120 seconds. It will result in a PlatformException.


@gildaswise I can verify it with `signInWithPhoneNumber(verificationId: null, smsCode: null)`, but I **can't send the sms code**, and **can't get auto-retrieved sms code**.

**Comment by gildaswise** [comment]
How can you verify the SMS code if you're passing null arguments? It should throw an error. Also, you can follow [this tutorial I made](https://medium.com/@gildaswise/flutter-adding-sign-in-with-google-and-phone-authentication-to-your-app-69f681518f9b), it sets up the phone verification on both Android and iOS but the problem of it overriding the previous signed in account persists, that's why I'm only using `signInWithGoogle` after the SMS code has been verified.

**Comment by kdy1** [comment]
@gildaswise It was just about signature. I'll edit it.

Only automatic retrieval was problematic, and I successfully used @creativecreatorormaybenot 's branch to link phone number to existing account. 

Maybe also title should be change to reflect the fact that instant verification and auto retrieval are problematic.

**Comment by awazgyawali** [comment]
This is a serious blocker to my project, I am letting the user sign in with google/facebook. After the AuthCredential is obtained from the 3rd party login I create a new user and  then I try to login with phone number and associate it to the current user, but the plugin automatically signs in with the new PhoneAuthCredential on this line https://github.com/flutter/plugins/blob/49c4ad9a079f2dbf79324f3957bc774a0e294970/packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java#L193

I want the raw PhoneAuthCredential as I would be able to sign in or link it to an existing user. Also I think it should be similar to what Google, Facebook sign-in work as they give cred object and we can use it as we like.

I would love to send a PR to fix this, please let me know if that's ok or maybe I am missing something.
Maybe @amirh can help.

**Comment by awazgyawali** [comment]
I have fixed this issue and landed a PR for it [here](https://github.com/flutter/plugins/pull/1568). Waiting for Googlers to review it.

**Comment by awazgyawali** [comment]
The PR has been merged. I think this issue should be closed as it has been merged. We are havving 5k+ issues and plenty of them are just not valid anymore. @collinjackson

**Comment by kroikie** [comment]
@kdy1

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/646. Any further collaboration will be done there.
